### ISSUE TITLE
Problem: clock_gettime broken on OSX + Homebrew

### ISSUE THREAD

**Issue by kozo2** [zeromq/libzmq#2119]
Sorry, in addition to #2116 , could you fix this problem too? https://github.com/Homebrew/homebrew-core/issues/5011#issuecomment-248228767
```
+    (buildpath/"src").install_symlink buildpath/"pollset.cpp"
+
+    cd "src" do
+      inreplace %w[clock.cpp clock.hpp condition_variable.hpp],
+        "clock_gettime", "my_clock_gettime"
+    end
```

The problem from ```(buildpath/"src").install_symlink buildpath/"pollset.cpp"``` was fixed in #2116 
but the other problems from the rest of the code have not been fixed yet

```
~> brew install czmq --HEAD
==> Using the sandbox
==> Cloning https://github.com/zeromq/czmq.git
Updating /Users/knishida/Library/Caches/Homebrew/czmq--git
==> Checking out branch master
==> ./autogen.sh
==> ./configure --prefix=/usr/local/Cellar/czmq/HEAD-50d948e_3 --with-libsodium
==> make
==> (ZSYS_INTERFACE=lo0 && make check-verbose)
Last 15 lines from /Users/knishida/Library/Logs/Homebrew/czmq/04.(ZSYS_INTERFACE=lo0:
2016-09-21 11:03:30 +0900

(ZSYS_INTERFACE=lo0 && make check-verbose)

/bin/sh ./libtool --mode=execute ./src/czmq_selftest -v
dyld: lazy symbol binding failed: Symbol not found: _clock_gettime
  Referenced from: /usr/local/opt/zeromq/lib/libzmq.5.dylib
  Expected in: /usr/lib/libSystem.B.dylib

dyld: Symbol not found: _clock_gettime
  Referenced from: /usr/local/opt/zeromq/lib/libzmq.5.dylib
  Expected in: /usr/lib/libSystem.B.dylib

make: *** [check-verbose] Trace/BPT trap: 5
```

**Comment by bluca** [comment]
Hi,

This works fine on the Travis OSX 10.9 VMs. Unfortunately I do not have an OSX platform to test it on.

If you check the generated src/platform.hpp, is ZMQ_HAVE_OSX defined?

OSX does not define clock_gettime, so it's reimplemented in src/clock.[h|c]pp if building on OSX.

**Comment by kozo2** [comment]
Hi,
I'm sorry I moved from OSX 10.11 to 10.12, and reinstalled homebrew.
Now I can not check the generated `src/platform.hpp`.

There is no error in 10.12 environment.
The terminal log like this.

```
~> brew install zeromq --HEAD
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
==> Updated Formulae
rswift

==> Installing dependencies for zeromq: autoconf, automake, libtool, pkg-config
==> Installing zeromq dependency: autoconf
==> Downloading https://homebrew.bintray.com/bottles/autoconf-2.69.sierra.bottle.4.tar.gz
######################################################################## 100.0%
==> Pouring autoconf-2.69.sierra.bottle.4.tar.gz
==> Caveats
Emacs Lisp files have been installed to:
  /usr/local/share/emacs/site-lisp/autoconf
==> Summary
🍺  /usr/local/Cellar/autoconf/2.69: 70 files, 3.0M
==> Installing zeromq dependency: automake
==> Downloading https://homebrew.bintray.com/bottles/automake-1.15.sierra.bottle.2.tar.gz
######################################################################## 100.0%
==> Pouring automake-1.15.sierra.bottle.2.tar.gz
🍺  /usr/local/Cellar/automake/1.15: 130 files, 2.9M
==> Installing zeromq dependency: libtool
==> Downloading https://homebrew.bintray.com/bottles/libtool-2.4.6_1.sierra.bottle.tar.gz
######################################################################## 100.0%
==> Pouring libtool-2.4.6_1.sierra.bottle.tar.gz
==> Caveats
In order to prevent conflicts with Apple's own libtool we have prepended a "g"
so, you have instead: glibtool and glibtoolize.
==> Summary
🍺  /usr/local/Cellar/libtool/2.4.6_1: 70 files, 3.7M
==> Installing zeromq dependency: pkg-config
==> Downloading https://homebrew.bintray.com/bottles/pkg-config-0.29.1_2.sierra.bottle.tar.gz
######################################################################## 100.0%
==> Pouring pkg-config-0.29.1_2.sierra.bottle.tar.gz
🍺  /usr/local/Cellar/pkg-config/0.29.1_2: 10 files, 627.4K
==> Installing zeromq
==> Using the sandbox
==> Cloning https://github.com/zeromq/libzmq.git
Cloning into '/Users/knishida/Library/Caches/Homebrew/zeromq--git'...
remote: Counting objects: 653, done.
remote: Compressing objects: 100% (567/567), done.
remote: Total 653 (delta 263), reused 172 (delta 47), pack-reused 0
Receiving objects: 100% (653/653), 742.72 KiB | 524.00 KiB/s, done.
Resolving deltas: 100% (263/263), done.
Checking connectivity... done.
==> Checking out branch master
==> ./autogen.sh
==> ./configure --prefix=/usr/local/Cellar/zeromq/HEAD-bacba21 --without-libsodium
==> make
==> make install
🍺  /usr/local/Cellar/zeromq/HEAD-bacba21: 11 files, 1.8M, built in 1 minute 14 seconds
~> brew install czmq --HEAD
==> Installing dependencies for czmq: libsodium
==> Installing czmq dependency: libsodium
==> Downloading https://homebrew.bintray.com/bottles/libsodium-1.0.11.sierra.bottle.tar.gz
######################################################################## 100.0%
==> Pouring libsodium-1.0.11.sierra.bottle.tar.gz
🍺  /usr/local/Cellar/libsodium/1.0.11: 66 files, 1.1M
==> Installing czmq
==> Using the sandbox
==> Cloning https://github.com/zeromq/czmq.git
Cloning into '/Users/knishida/Library/Caches/Homebrew/czmq--git'...
remote: Counting objects: 748, done.
remote: Compressing objects: 100% (640/640), done.
remote: Total 748 (delta 247), reused 202 (delta 70), pack-reused 0
Receiving objects: 100% (748/748), 1.13 MiB | 369.00 KiB/s, done.
Resolving deltas: 100% (247/247), done.
Checking connectivity... done.
==> Checking out branch master
==> ./autogen.sh
==> ./configure --prefix=/usr/local/Cellar/czmq/HEAD-50d948e_3 --with-libsodium
==> make
==> (ZSYS_INTERFACE=lo0 && make check-verbose)
==> make install
🍺  /usr/local/Cellar/czmq/HEAD-50d948e_3: 84 files, 1.9M, built in 1 minute 23 seconds

```

I close this issue once.
Sorry I couldn't help you.

**Comment by jimrthy** [comment]
I ran into what looks like the same error on OSX 10.11.6, except that I'm going through the build process steps by hand instead of using homebrew.

platform.hpp definitely does define ZMQ_HAVE_OSX.

It looks like the check for __MAC_OS_X_VERSION_MIN_REQUIRED < 101200 around the clock_gettime calls is failing.

I didn't see that macro defined anywhere, so I tried defining it as part of the CFLAGS to configure. That didn't help.

So I tried defining it manually at the top of clock.cpp and got a warning about redefining it, along with its source file.

The original definition is based on a macro that, according to some comments, gets set by the compiler.

It looks like I have a newer version of Xcode that includes the 10.12 SDK.

Anyway. That symbol redef took me from 85 failing tests to 8.

I had to repeat this at the top of condition_variable.hpp to get them all to pass.

This is pretty obviously an ugly hack, but it seems like a viable work-around. I'm not sure what a good solution would look like.
